### PR TITLE
sub-repo commit filtering respects requested number of commits

### PR DESCRIPTION
**Pull request by mollylogue**
if sub-repo filtering filters out commits, attempt to fetch the requested number of commits

We may need to fetch more commits after filtering. To limit the impact of this we'll try
to fetch more repos up to a maximum of 5 times before giving up and returning what we have.

Closes https://github.com/sourcegraph/sourcegraph/issues/30750

### CODE REVIEW

File: internal/vcs/git/commits.go [hunk 1]
```diff
@@ -285,6 +285,19 @@ func isBadObjectErr(output, obj string) bool {
 //
 // The caller is responsible for doing checkSpecArgSafety on opt.Head and opt.Base.
 func commitLog(ctx context.Context, repo api.RepoName, opt CommitsOptions, checker authz.SubRepoPermissionChecker) (commits []*gitdomain.Commit, err error) {
+	wrappedCommits, err := getWrappedCommits(ctx, repo, opt)
+	if err != nil {
+		return nil, err
+	}
+	filtered, err := filterCommits(ctx, wrappedCommits, repo, checker)
```
Comment (ryanslade) [inline internal/vcs/git/commits.go:292]: We should fail early here if `err` is not nil.

**Comment by ryanslade** [comment]
Approach looks good to me. I think you're correct to worry about worst case performance but maybe we can get around that by either just picking an upper limit on the number of times we loop around in `getMoreCommits` or add a timeout of say 5 seconds and just return whatever we have at that point.

In general this part of the code is quite slow as it is. Even without doing any filtering the code is quite naive when you click "show more" as it just increases the count each time and requests everything again so this could be something to look into in future.

**Comment by ryanslade** [comment]
I'm going to take this over as @mollylogue is on support this week.

**Comment by ryanslade** [comment]
@mollylogue I think I've got this to a point where could merge, PTAL

**Comment by mollylogue** [comment]
> @mollylogue I think I've got this to a point where could merge, PTAL

@ryanslade Thanks so much for doing this! Sorry, I probably should've more formally handed this off to you. Naively I thought maybe I'd have some free time to get it across the line this week 😅

**Pull request merged**